File: leetcode/Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.java
import java.util.*;
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    int[] inorder;
    int[] postorder;
    HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
    
    TreeNode buildTree(int ia, int ib, int pa, int pb) {
        if (ia > ib || pa > pb) return null;
        TreeNode r = new TreeNode(postorder[pb]);
        int m = map.get(r.val);
        int l = m - ia;
        r.left = buildTree(ia, m - 1, pa, pa + l - 1);
        r.right = buildTree(m + 1, ib, pa + l, pb - 1);
        return r;
    }
    
    public TreeNode buildTree(int[] __inorder, int[] __postorder) {
        inorder = __inorder;
        postorder = __postorder;
        for (int i = 0; i < inorder.length; ++i) {
            map.put(inorder[i], i);
        }
        return buildTree(0, inorder.length - 1, 0, postorder.length - 1);
    }
}

File: leetcode/Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    TreeNode *build(vector<int> &inorder,   int begin_in,   int end_in, 
                    vector<int> &postorder, int begin_post, int end_post) {
        // Assume the input is valid.
        if (begin_in >= end_in)
            return NULL;
        vector<int>::iterator iter_mid = find(inorder.begin() + begin_in,
                                              inorder.begin() + end_in,
                                              postorder[end_post - 1]);
        int mid = iter_mid - inorder.begin();
        TreeNode *node = new TreeNode(inorder[mid]);
        node->left  = build(
                        inorder,   begin_in,   mid,
                        postorder, begin_post, begin_post + (mid - begin_in));
        node->right = build(
                        inorder,   mid + 1,                       end_in,
                        postorder, begin_post + (mid - begin_in), end_post - 1);
        return node;
    }
public:
    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
        return build(inorder,   0, inorder.size(),
                     postorder, 0, postorder.size());
    }
};
